🐛 BUG: c3 project name input is not correctly validated

### Which Cloudflare product(s) does this pertain to?

C3

### What version of `Wrangler` are you using?

c3@2.0.0-next.2

### What operating system are you using?

Mac

### Describe the Bug

If I specify `foo.com` as my project name in C3, C3 happily accepts it just to fail during deployment:

```
npm create cloudflare@next     

using create-cloudflare version 2.0.0-next.1

╭ Create an application with Cloudflare Step 1 of 3
│ 
├ Where do you want to create your application?
│ dir ilovejavascript.net
│
├ What type of application do you want to create?
│ type webFramework
│
├ Which development framework do you want to use?
│ framework Angular
│
╰ Continue with Angular Creating your application via `npx @angular/cli@next new ilovejavascript.net --standalone`

? Would you like to add Angular routing? No
? Which stylesheet format would you like to use? CSS
            Standalone application structure is new and not yet supported by many existing 'ng add' and 'ng update' integrations with community libraries.
CREATE ilovejavascript.net/README.md (1072 bytes)
CREATE ilovejavascript.net/.editorconfig (274 bytes)
CREATE ilovejavascript.net/.gitignore (548 bytes)
CREATE ilovejavascript.net/angular.json (3015 bytes)
CREATE ilovejavascript.net/package.json (1050 bytes)
CREATE ilovejavascript.net/tsconfig.json (901 bytes)
CREATE ilovejavascript.net/tsconfig.app.json (263 bytes)
CREATE ilovejavascript.net/tsconfig.spec.json (273 bytes)
CREATE ilovejavascript.net/.vscode/extensions.json (130 bytes)
CREATE ilovejavascript.net/.vscode/launch.json (470 bytes)
CREATE ilovejavascript.net/.vscode/tasks.json (938 bytes)
CREATE ilovejavascript.net/src/main.ts (250 bytes)
CREATE ilovejavascript.net/src/favicon.ico (1642 bytes)
CREATE ilovejavascript.net/src/index.html (304 bytes)
CREATE ilovejavascript.net/src/styles.css (80 bytes)
CREATE ilovejavascript.net/src/app/app.component.css (0 bytes)
CREATE ilovejavascript.net/src/app/app.component.html (23083 bytes)
CREATE ilovejavascript.net/src/app/app.component.spec.ts (927 bytes)
CREATE ilovejavascript.net/src/app/app.component.ts (318 bytes)
CREATE ilovejavascript.net/src/app/app.config.ts (117 bytes)
CREATE ilovejavascript.net/src/assets/.gitkeep (0 bytes)
✔ Packages installed successfully.
    Successfully initialized git.
╭ Configuring your application for Cloudflare Step 2 of 3
│ 
├ Installing Angular SSR 
│ installed
│ 
├ Adding Cloudflare Pages adaptor code 
│ added Cloudflare Pages adaptor code`
│ 
├ Installing adaptor dependencies 
│ Installed
│ 
├ Updating application code 
│ Done updating application code
│ 
├ Updating angular.json config 
│ Updated angular.json config
│ 
├ Adding command scripts for development and deployment
│ added commands to `package.json`
│ 
╰ Application configured 

╭ Deploy with Cloudflare Step 3 of 3
│ 
├ Do you want to deploy your application?
│ yes deploying via `npm run deploy`
│
┐ Creating Pages project 

✘ [ERROR] A request to the Cloudflare API (/accounts/d61d1e4d1ab4db4f5dbcf0a11cf90a5e/pages/projects) failed.

  You have entered an invalid project name. Project names can be 1-58 lowercase characters with dashes, but cannot start/end with a dash (/). [code: 8000003]
  
  If you think this is a bug, please open an issue at: https://github.com/cloudflare/workers-sdk/issues/new/choose
````

In the short term we should validate the C3 input.

In the mid-term/long-term we should reconsider what kind of validation constraints we want to put on the project name as `foo.com` project names that match the target domain seem like a perfectly valid choice we should support somehow (e.g. by converting the name to foo-com when deploying onto `.pages.dev` subdomain).

--- a/.changeset/purple-eels-own.md
+++ b/.changeset/purple-eels-own.md
@@ -1,0 +1,5 @@
+---
+"create-cloudflare": patch
+---
+
+Verify that project names are valid for pages projects

--- a/packages/create-cloudflare/e2e-tests/workers.test.ts
+++ b/packages/create-cloudflare/e2e-tests/workers.test.ts
@@ -27,7 +27,7 @@ describe.skipIf(frameworkToTest || isQuarantineMode())(
 		});
 
 		const runCli = async (template: string) => {
-			const projectPath = getPath(template);
+			const projectPath = getPath(template.toLowerCase());
 
 			const argv = [
 				projectPath,

--- a/packages/create-cloudflare/src/__tests__/common.test.ts
+++ b/packages/create-cloudflare/src/__tests__/common.test.ts
@@ -1,6 +1,7 @@
 import * as command from "helpers/command";
 import { describe, expect, test, vi } from "vitest";
 import { isGitConfigured } from "../common";
+import { validateProjectDirectory } from "../common";
 
 function promisify<T>(value: T) {
 	return new Promise<T>((res) => res(value));
@@ -37,5 +38,27 @@ describe("isGitConfigured", () => {
 			throw new Error("git not found");
 		});
 		expect(await isGitConfigured()).toBe(false);
+	});
+});
+
+describe("validateProjectDirectory", () => {
+	test("allow valid project names", async () => {
+		expect(validateProjectDirectory("foo")).toBeUndefined();
+		expect(validateProjectDirectory("foo/bar/baz")).toBeUndefined();
+		expect(validateProjectDirectory("./foobar")).toBeUndefined();
+		expect(validateProjectDirectory("f".repeat(58))).toBeUndefined();
+	});
+
+	test("disallow invalid project names", async () => {
+		// Invalid pages project names should return an error
+		expect(validateProjectDirectory("foobar-")).not.toBeUndefined();
+		expect(validateProjectDirectory("-foobar-")).not.toBeUndefined();
+		expect(validateProjectDirectory("fo*o{ba)r")).not.toBeUndefined();
+		expect(validateProjectDirectory("f".repeat(59))).not.toBeUndefined();
+	});
+
+	test("disallow existing, non-empty directories", async () => {
+		// Existing, non-empty directories should return an error
+		expect(validateProjectDirectory(".")).not.toBeUndefined();
 	});
 });

--- a/packages/create-cloudflare/src/common.ts
+++ b/packages/create-cloudflare/src/common.ts
@@ -45,7 +45,7 @@ export const validateProjectDirectory = (relativePath: string) => {
 
 	// Ensure the name is valid per the pages schema
 	const projectName = basename(path);
-	const invalidChars = /^- | -$|[^a-z0-9-]/;
+	const invalidChars = /[^a-z0-9-]/;
 	const invalidStartEnd = /^-|-$/;
 
 	if (projectName.match(invalidStartEnd)) {

--- a/packages/create-cloudflare/src/common.ts
+++ b/packages/create-cloudflare/src/common.ts
@@ -34,8 +34,17 @@ import type { C3Args, PagesGeneratorContext } from "types";
 const { name, npm } = detectPackageManager();
 
 export const validateProjectDirectory = (relativePath: string) => {
+	// Validate that the directory is non-existant or empty
+	const path = resolve(relativePath);
+	const existsAlready = existsSync(path);
+	const isEmpty = existsAlready && readdirSync(path).length === 0; // allow existing dirs _if empty_ to ensure c3 is non-destructive
+
+	if (existsAlready && !isEmpty) {
+		return `Directory \`${relativePath}\` already exists and is not empty. Please choose a new name.`;
+	}
+
 	// Ensure the name is valid per the pages schema
-	const projectName = basename(relativePath);
+	const projectName = basename(path);
 	const invalidChars = /^- | -$|[^a-z0-9-]/;
 	const invalidStartEnd = /^-|-$/;
 
@@ -49,15 +58,6 @@ export const validateProjectDirectory = (relativePath: string) => {
 
 	if (projectName.length > 58) {
 		return `Project names must be less than 58 characters.`;
-	}
-
-	// Validate that the directory is non-existant or empty
-	const path = resolve(relativePath);
-	const existsAlready = existsSync(path);
-	const isEmpty = existsAlready && readdirSync(path).length === 0; // allow existing dirs _if empty_ to ensure c3 is non-destructive
-
-	if (existsAlready && !isEmpty) {
-		return `Directory \`${relativePath}\` already exists and is not empty. Please choose a new name.`;
 	}
 };
 

--- a/packages/create-cloudflare/src/common.ts
+++ b/packages/create-cloudflare/src/common.ts
@@ -34,6 +34,24 @@ import type { C3Args, PagesGeneratorContext } from "types";
 const { name, npm } = detectPackageManager();
 
 export const validateProjectDirectory = (relativePath: string) => {
+	// Ensure the name is valid per the pages schema
+	const projectName = basename(relativePath);
+	const invalidChars = /^- | -$|[^a-z0-9-]/;
+	const invalidStartEnd = /^-|-$/;
+
+	if (projectName.match(invalidStartEnd)) {
+		return `Project name cannot start or end with a dash.`;
+	}
+
+	if (projectName.match(invalidChars)) {
+		return `Project name must only contain lowercase characters, numbers, and dashes.`;
+	}
+
+	if (projectName.length > 58) {
+		return `Project names must be less than 58 characters.`;
+	}
+
+	// Validate that the directory is non-existant or empty
 	const path = resolve(relativePath);
 	const existsAlready = existsSync(path);
 	const isEmpty = existsAlready && readdirSync(path).length === 0; // allow existing dirs _if empty_ to ensure c3 is non-destructive

